miurla/morphic
# Local redis support

# Add Local Redis Support

## Motivation
This PR introduces support for local Redis alongside the existing Upstash Redis implementation. This addition aims to enhance development workflows by allowing developers to work with a local Redis instance, reducing dependency on external services during development and testing.

## Summary of Changes
- Implemented a `RedisWrapper` class to abstract Redis operations, supporting both Upstash and local Redis clients.
- Added configuration options to switch between local and Upstash Redis using the `USE_LOCAL_REDIS` environment variable.
- Updated Redis-related functions (`getChats`, `saveChat`, etc.) to work with the new `RedisWrapper`.
- Enhanced error handling and logging for Redis operations to improve debugging and reliability.
- Updated documentation to reflect the new local Redis option.

## Testing Instructions
1. Clone the updated repository and install dependencies.
2. Set up a local Redis instance (instructions in README.md).
3. Copy `.env.example` to `.env` and set `USE_LOCAL_REDIS=true`.
4. Modify docker-compose.yaml to use Dockerfile.test instead of Dockerfile
5. Run the application and perform the following tests:
   - Create a new chat/search
   - Verify that the chat history is saved and can be retrieved
   - Perform multiple searches and ensure chat history persists
   - Clear chat history and verify it's removed from Redis
6. Set `USE_LOCAL_REDIS=false` and repeat steps 4-5 to verify Upstash Redis still works as expected.

## Potential Impacts
- No breaking changes to existing functionality.
- Improved development workflow with local Redis option.
- Slight changes to Redis key structures for better compatibility between local and Upstash Redis.

## Additional Notes
- This PR does not modify any core application logic, focusing solely on the Redis implementation.
- Performance impact should be minimal, with potential improvements for local development environments.
- Future considerations may include adding more Redis operations as needed and further optimizing the `RedisWrapper` class.

## Checklist
- [x] Updated README.md with local Redis setup instructions
- [x] Updated .env.example with new environment variables
- [x] Performed manual testing with both local and Upstash Redis
- [x ] Removed unnecessary debug logging used for testing
- [x] Reviewed and updated code comments

Closes #[110] Provide local Redis option
